joeynmt/datasets.py

```diff
@@ -106,6 +107,9 @@ def trg(self) -> List[str]:
         """get detokenized preprocessed data in trg language."""
         return self.get_list(self.trg_lang, tokenized=False) if self.has_trg else []
 
+    def make_iter(self, **kwargs):
+        return make_data_iter(dataset=self, **kwargs)
```

Comment:
My custom solution for my continuous case would then be:

1. Define a custom batch class
```python
class SignBatch(Batch):
    def __init__(self, src_length, **kwargs):
        super().__init__(src_length=src_length, **kwargs)
        self.src_mask = self._pose_mask(src_length)

    def _pose_mask(self, pose_length: torch.Tensor) -> torch.Tensor:
        max_len = pose_length.max().item()
        return torch.arange(max_len)[None, :] < pose_length[:, None]
```

2. Define a custom collation function  to `zero_pad` and instantiate the batch:

```python
def collate_fn(
        batch: List[Dict[str, Any]],
        pad_index: int = PAD_ID,
        device: torch.device = CPU_DEVICE,
        has_trg: bool = True,
        is_train: bool = True) -> Batch:
    collated = zero_pad_collator(batch)

    return SignBatch(
        src=collated["src"],
        src_length=collated["src_length"],
        trg=collated["trg"],
        trg_length=collated["trg_length"],
        device=device,
        pad_index=pad_index,
        has_trg=has_trg,
        is_train=is_train,
    )
```
3. Overriding the dataset `make_iter` method:
```python
def make_iter(self, pad_index: int = PAD_ID,
                  device: torch.device = CPU_DEVICE,
                  **kwargs) -> DataLoader:
    data_loader = super().make_iter(**kwargs)
    dataset = data_loader.dataset

    # data iterator
    return DataLoader(
        dataset=dataset,
        batch_sampler=data_loader.batch_sampler,
        collate_fn=partial(
            collate_fn,
            pad_index=pad_index,
            device=device,
            has_trg=True,
            is_train=dataset.split == "train",
        ),
        num_workers=data_loader.num_workers,
    )
```

Reply:
thanks for this elaboration @AmitMY, this is very helpful.